### PR TITLE
:sparkles: Enable regression & misc

### PR DESCRIPTION
**Pull request by o-laurent**
- [x] Compress GNNL and NNL
- [x] Make the MLP more flexible
- [x] Do we really need to add a parameter to the CLI (regression vs. classification)? Why not use dm for this?
- [x] Add VGGs
- [x] Add LeNet
- [x] Add a setup.py when poetry is not available + mention it on the installation page
- [x] Improve the MLP baseline
- [x] Add ensemble regression
- [x] How to deal with mutlivariate regression cases? (different output) -> Ask for the dimension first
- [x] Fix Bayesian layers? (cf failed tests: #28)
- [x] Add bayesian models & losses
- [x] Add deep ensembles method
- [x] Add a Bayesian tutorial
- [ ] Fix tutorial generation
- [ ] Make tutorial gallery static
- [ ] Add a pretrained tutorial
- [ ] Add a regression tutorial
- [ ] Add tests
- [ ] Add docs
- [ ] Improve VGG optim recipes?

Done for now:
Add an arg to CLI to split regression and cls :hammer: Modify tests accordingly :heavy_check_mark: 
Add a regression recipe :sparkles:
Add a small MLP model :sparkles:
Add an MLP baseline :sparkles:
Add a UCIRegression Datamodule :sparkles:
Add GaussianNLL lightning metric :sparkles:
Add an experiment :sparkles:

### CODE REVIEW

File: torch_uncertainty/metrics/nll.py [hunk 1]
```diff
@@ -94,3 +94,95 @@ def compute(self) -> torch.Tensor:
             return values.sum(dim=-1) / self.total
         else:  # reduction is None or "none"
             return values
+
+
+class GaussianNegativeLogLikelihood(Metric):
```
Comment (o-laurent) [inline torch_uncertainty/metrics/nll.py:99]: Compress the code

**Comment by o-laurent** [comment]
Documentation is missing the code seems approx fine.

File: experiments/classification/mnist/bayesian_lenet.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    Uses Adam default hyperparameters.
+
+    Args:
+        model (nn.Module): LeNet model.
+    """
+    optimizer = optim.Adam(
+        model.parameters(),
+        lr=1e-3,
+    )
+    # scheduler = optim.lr_scheduler.CosineAnnealingLR(
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply experiments/classification/mnist/bayesian_lenet.py:29]: (and spaces)

File: torch_uncertainty/models/vgg/packed.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    "packed_vgg16",
+    "packed_vgg19",
+]
+
+
+def packed_vgg11(
+    in_channels: int,
+    num_classes: int,
+    alpha: int,
+    num_estimators: int,
```
Comment (o-laurent) [inline torch_uncertainty/models/vgg/packed.py:23]: check argument order

File: torch_uncertainty/models/utils.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from ..layers.bayesian_layers import bayesian_modules
+
+
+def StochasticModel(Model):
+    """Decorator for stochastic models. When applied to a model, it adds the
+    freeze and unfreeze methods to the model. Use freeze to freeze the
```
Comment (o-laurent) [inline torch_uncertainty/models/utils.py:6]: verbose

File: torch_uncertainty/models/lenet.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        activation=activation,
+        groups=groups,
+        dropout=dropout,
+    )
+
+
+def bayesian_lenet(
+    in_channels: int,
+    num_classes: int,
+    # prior_mu: float = 0.0,
```
Comment (o-laurent) [inline torch_uncertainty/models/lenet.py:160]: Put model default values and remove comments

File: torch_uncertainty/metrics/nll.py [hunk 6]
```diff
@@ -94,3 +94,54 @@ def compute(self) -> torch.Tensor:
...
+            - ``'mean'`` [default]: Averages score across samples
+            - ``'sum'``: Sum score across samples
+            - ``'none'`` or ``None``: Returns score per sample
+
+        kwargs: Additional keyword arguments, see `Advanced metric settings
+            <https://torchmetrics.readthedocs.io/en/stable/pages/overview.html#metric-kwargs>`_.
+
+    Inputs:
+        - :attr:`probs`: :math:`(B, C)`
+        - :attr:`target`: :math:`(B)`
```
Comment (o-laurent) [inline torch_uncertainty/metrics/nll.py:115]: var is missing

File: torch_uncertainty/layers/bayesian_layers/sampler.py [hunk 7]
```diff
@@ -7,26 +7,27 @@
 
 # fmt: on
 class TrainableDistribution(nn.Module):
-    lsqrt2pi = np.log(np.sqrt(2 * np.pi))
+    lsqrt2pi = torch.as_tensor(np.log(np.sqrt(2 * np.pi)))
```
Comment (o-laurent) [inline torch_uncertainty/layers/bayesian_layers/sampler.py:10]: Try to remove numpy?

File: torch_uncertainty/layers/bayesian_layers/bayes_linear.py [hunk 8]
```diff
@@ -21,7 +21,7 @@ class BayesLinear(nn.Module):
         mu_init (float, optional): Initial mean of the posterior distribution.
             Defaults to 0.0.
         sigma_init (float, optional): Initial standard deviation of the
-            posterior distribution. Defaults to 10.0.
+            posterior distribution. Defaults to 2.0.
```
Comment (o-laurent) [inline torch_uncertainty/layers/bayesian_layers/bayes_linear.py:24]: fix default value in docstring

File: torch_uncertainty/datasets/uci_regression.py [hunk 9]
```diff
@@ -231,22 +276,43 @@ def _make_dataset(self) -> None:
...
 
     def __getitem__(self, index: int) -> Tuple[torch.Tensor, torch.Tensor]:
         """Get sample and target for a given index."""
-        data = self.data[index]
-        if self.transform is not None:
-            data = self.transform(data)
-        target = self.targets[index]
-        if self.target_transform is not None:
-            target = self.target_transform(target)
+        if self.dataset_name == "energy-prediction":
```
Comment (o-laurent) [inline torch_uncertainty/datasets/uci_regression.py:306]: remove energy predictions special code lines

File: torch_uncertainty/baselines/regression/mlp.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        num_outputs: int,
+        in_features: int,
+        loss: nn.Module,
+        optimization_procedure: Any,
+        version: Literal["vanilla", "packed"],
+        hidden_dims: List[int],
+        dist_estimation: bool,
+        num_estimators: Optional[int] = 1,
+        alpha: Optional[float] = None,
+        gamma: Optional[int] = 1,
```
Comment (o-laurent) [inline torch_uncertainty/baselines/regression/mlp.py:36]: Remove optional

File: torch_uncertainty/datamodules/cifar10.py [hunk 11]
```diff
@@ -117,7 +142,11 @@ def setup(self, stage: Optional[str] = None) -> None:
                 transform=self.transform_train,
             )
             self.train, self.val = random_split(
-                full, [len(full) - self.val_split, self.val_split]
+                full,
+                [
+                    int(len(full) ** (1 - self.val_split)),
```
Comment (alafage) [inline torch_uncertainty/datamodules/cifar10.py:147]: Why the double ** here? Shouldn't it be simply * instead?
Other datamodules have the same implementation

**Review by alafage — CHANGES_REQUESTED** [review]
It seems to me that there are typos in some datamodules when defining both train and validation sets using random_split. Otherwise this pull request is fine enough to be merged to dev :)

File: torch_uncertainty/datamodules/cifar10.py (hunk 11 above)
Comment (o-laurent) [reply torch_uncertainty/datamodules/cifar10.py:147]: You're right! Changing it now
